README.md
# Native SSE (EventSource) Plugin for Capacitor

Because of the way Capacitor plugins are designed, it does not exactly follow the format/signature of the "real" browser `EventSource` API, but it works similarly.

## Installation

```shell
npm install capacitor-eventsource
npx cap sync
```

On iOS or for pure-web use, no further steps are needed.

### Android Installation

On Android, update your main activity to pull in the `EventSource` plugin:

```java
import com.raccoonfink.plugins.eventsource.EventSource;

public class MainActivity extends BridgeActivity {
  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    // Initializes the Bridge
    this.init(savedInstanceState, new ArrayList<Class<? extends Plugin>>() {{
      // Additional plugins you've installed go here
      // Ex: add(TotallyAwesomePlugin.class);
      add(EventSource.class);
    }});
  }
}
```

## Usage

```javascript
EventSource.configure({
  url: 'https://my-sse-url/',
  reconnectTime: 1000,
  maxReconnectTime: 60000,
  idleTimeout: 30000,
});
EventSource.addListener('open', (openResult) => {
  console.debug('Connection opened.');
});
EventSource.addListener('message', (messageResult) => {
  console.info('I got a message: ' + messageResult.message);
});
EventSource.addListener('error', (errorResult) => {
  console.error('Uh-oh: ' + errorResult.error);
});
EventSource.addListener('readyStateChanged', (readyStateChangedResult) => {
  console.debug('Ready state changed: ' + readyStateChangedResult.state);
});
```

## Parameters

*note*: all times are in milliseconds

* url: **required** the URL to connect to
* reconnectTime: time to reconnect after the first failure (iOS/Android)
* maxReconnectTime: maximum amount of time to wait until attempting a reconnect (iOS/Android)
* idleTimeout: maximum amount of time between receiving any data before considering the connection to have timed out (iOS/Android)


CHANGELOG.md
Changes
=======

0.1.10
-----
* Add some defensive code to the Android version to avoid crashes.

0.1.9
-----
* Fix cleanup after closing in the pure-web version.

0.1.8
-----
* Don't retain messages, just let them drop.

0.1.7
-----
* Wait!  No!  It was something different!  It's aliiiiive!

0.1.6
-----
* Sigh, the launchdarkly one crashes.  :(

0.1.5
-----
* More fixes for potential race conditions.
* Fix a compilation/import thing on Android.

0.1.4
-----

* Dependency updates.
* Properly export the new configuration parameters

0.1.3
-----

* Update Android implementation to use launchdarkly (haha I know)
* Pass parameters for configuring retries/timeout/etc

0.1.2
-----

* Update Android implementation to use okhttp-sse
* Skip updates that come through after shutting down, to avoid
  a potential crash in notifying listeners

0.1.1
-----

* Dependency updates.

0.1.0
-----

* More iOS switching! Now using LDSwiftEventSource ;)

0.0.9
-----

* Fix typo in Android build.

0.0.7
-----

* Add typescript definition for `removeAllListeners()`

0.0.6
-----

* Always close, just in case.

0.0.5
-----

* Put it back.  :)

0.0.3/0.0.4
-----------

* Switch to IKEventSource on iOS.

0.0.2
-----

* Fix worker threads not communicating properly on Android.

0.0.1
-----

* Initial implementation.


src/web.ts
import { WebPlugin } from '@capacitor/core';
import {
  EventSourceOptions,
  EventSourcePlugin,
  READY_STATE,
} from './definitions';

export class EventSourceWeb extends WebPlugin implements EventSourcePlugin {
  private url?: string;
  private eventSource?: EventSource | null;
  private opened = false;

  constructor() {
    super({
      name: 'EventSource',
      platforms: ['web'],
    });
  }

  async configure(options: EventSourceOptions): Promise<void> {
    console.debug(`EventSourceWeb.configure(${options.url})`);
    if (options.url) {
      this.url = options.url;
    } else {
      throw new Error('url is required');
    }
  }

  async open(): Promise<void> {
    console.debug(`EventSourceWeb.open()`);
    if (!this.url) {
      throw new Error('You must call configure first!');
    }

    this.notifyListeners('readyStateChanged', {
      state: READY_STATE.CONNECTING,
    });

    this.eventSource = new window.EventSource(this.url);
    this.eventSource.onopen = this.onOpen.bind(this);
    this.eventSource.onmessage = this.onMessage.bind(this);
    this.eventSource.onerror = this.onError.bind(this);
  }

  onError(ev: any) {
    this.notifyListeners('error', { error: ev?.message });
  }

  onMessage(ev: MessageEvent) {
    if (this.opened) {
      this.notifyListeners('message', { message: ev?.data });
    }
  }

  onOpen(ev: any) {
    this.opened = true;
    this.notifyListeners('open', { message: ev?.message });
    this.notifyListeners('readyStateChanged', { state: READY_STATE.OPEN });
  }

  async close(): Promise<void> {
    console.debug(`EventSourceWeb.close()`);
    this.opened = false;
    if (this.eventSource) {
      this.eventSource.close();
      this.eventSource.onopen = null;
      this.eventSource.onmessage = null;
      this.eventSource.onerror = null;
      this.eventSource = null;
      this.notifyListeners('readyStateChanged', {
        state: READY_STATE.CLOSED,
      });
    }
  }
}

const EventSourcePluginImpl = new EventSourceWeb();

export { EventSourcePluginImpl as EventSource };

import { registerWebPlugin } from '@capacitor/core';
registerWebPlugin(EventSourcePluginImpl);


src/definitions.ts
import { PluginListenerHandle } from '@capacitor/core';

declare module '@capacitor/core' {
  interface PluginRegistry {
    EventSource: EventSourcePlugin;
  }
}

export interface EventSourceOptions {
  url: string;
  reconnectTime?: number;
  maxReconnectTime?: number;
  backoffResetThreshold?: number;
  idleTimeout?: number;
}

export interface OpenResult {
  value?: string;
}

export interface MessageResult {
  message?: string;
}

export interface ErrorResult {
  error?: string;
}

export enum READY_STATE {
  CONNECTING = 0,
  OPEN = 1,
  CLOSED = 2,
}

export interface ReadyStateChangedResult {
  state?: number;
}

export interface EventSourcePlugin {
  configure(options: EventSourceOptions): Promise<void>;
  open(): Promise<void>;
  close(): Promise<void>;

  removeAllListeners(): void;
  addListener(
    eventName: 'open',
    listenerFunc: (result: OpenResult) => void,
  ): PluginListenerHandle;
  addListener(
    eventName: 'message',
    listenerFunc: (result: MessageResult) => void,
  ): PluginListenerHandle;
  addListener(
    eventName: 'error',
    listenerFunc: (result: ErrorResult) => void,
  ): PluginListenerHandle;
  addListener(
    eventName: 'readyStateChanged',
    listenerFunc: (result: ReadyStateChangedResult) => void,
  ): PluginListenerHandle;
}


ios/Plugin/Plugin.swift
import Foundation
import Capacitor
import LDSwiftEventSource

#if !os(Linux)
import os.log
#endif

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(EventSource)
public class EventSource: CAPPlugin, LDSwiftEventSource.EventHandler {
    #if !os(Linux)
    private let logger: OSLog = OSLog(subsystem: "com.raccoonfink.capacitor.eventsource", category: "EventSource")
    #endif

    var eventSource: LDSwiftEventSource.EventSource?
    var url: String?
    var opened = false

    @objc func configure(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("you must provide a URL")
            return
        }

        #if !os(Linux)
        os_log("configuring event source with URL %s", log: logger, type: .info, url)
        #endif

        guard let serverURL = URL.init(string: url) else {
            call.reject("invalid URL")
            return
        }
        self.url = url
        if self.eventSource != nil && self.opened {
            #if !os(Linux)
            os_log("configure() called: closing existing event source", log: logger, type: .error)
            #endif
            self.opened = false
            self.eventSource?.stop()
        }
        var config = LDSwiftEventSource.EventSource.Config(handler: self, url: serverURL)

        config.reconnectTime         = (call.getDouble("reconnectTime")         ?? 1000.0)  / 1000.0
        config.maxReconnectTime      = (call.getDouble("maxReconnectTime")      ?? 60000.0) / 1000.0
        config.backoffResetThreshold = (call.getDouble("backoffResetThreshold") ?? 5000.0)  / 1000.0
        config.idleTimeout           = (call.getDouble("idleTimeout")           ?? 30000.0) / 1000.0

        self.eventSource = LDSwiftEventSource.EventSource.init(config: config)
        call.resolve()
    }

    @objc func open(_ call: CAPPluginCall) {
        if self.eventSource == nil {
            call.reject("you must configure the event source first")
            return
        }

        #if !os(Linux)
        os_log("opening event source to %s", log: logger, type: .info, self.url ?? "unknown")
        #endif

        self.notifyListeners("readyStateChanged", data: [
            "state": ReadyState.connecting as Any
        ], retainUntilConsumed: false)

        self.opened = true
        self.eventSource?.start()
        call.resolve()
    }

    @objc func close(_ call: CAPPluginCall) {
        #if !os(Linux)
        os_log("closing event source connection to %s", log: logger, type: .info, self.url ?? "unknown")
        #endif

        if self.eventSource != nil {
            self.opened = false
            self.eventSource?.stop()
            self.eventSource = nil
        }
        call.resolve()
    }

    public func onOpened() {
        if !self.opened {
            #if !os(Linux)
            os_log("onOpened skipped (self.opened=false)", log: logger, type: .debug)
            #endif
            return
        }

        self.notifyListeners("readyStateChanged", data: [
            "state": ReadyState.open as Any
        ], retainUntilConsumed: false)
        self.notifyListeners("open", data: [String: Any](), retainUntilConsumed: false)
    }

    public func onClosed() {
        if !self.opened {
            #if !os(Linux)
            os_log("onClosed skipped (self.opened=false)", log: logger, type: .debug)
            #endif
            return
        }

        self.notifyListeners("readyStateChanged", data: [
            "state": ReadyState.closed as Any
        ], retainUntilConsumed: false)
    }

    public func onMessage(eventType: String, messageEvent: MessageEvent) {
        if !self.opened {
            #if !os(Linux)
            os_log("onMessage skipped (self.opened=false)", log: logger, type: .debug)
            #endif
            return
        }

        self.notifyListeners("message", data: [
            "type": eventType,
            "message": messageEvent.data as Any
        ], retainUntilConsumed: false)
    }

    public func onComment(comment: String) {
        #if !os(Linux)
        os_log("comment ignored: %s", log: logger, type: .debug, comment)
        #endif
    }

    public func onError(error: Error) {
        if !self.opened {
            #if !os(Linux)
            os_log("onError skipped (self.opened=false, error=%s)", log: logger, type: .debug, error.localizedDescription)
            #endif
            return
        }

        self.notifyListeners("error", data: [
            "error": error.localizedDescription as Any
        ], retainUntilConsumed: false)
    }

}


ios/PluginTests/PluginTests.swift
import XCTest
import Capacitor
@testable import Plugin

class PluginTests: XCTestCase {

    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }

    /*
    func testEcho() {
        // This is an example of a functional test case for a plugin.
        // Use XCTAssert and related functions to verify your tests produce the correct results.

        let value = "Hello, World!"
        let plugin = MyPlugin()

        let call = CAPPluginCall(callbackId: "test", options: [
            "value": value
        ], success: { (result, _) in
            let resultValue = result!.data["value"] as? String
            XCTAssertEqual(value, resultValue)
        }, error: { (_) in
            XCTFail("Error shouldn't have been called")
        })

        plugin.echo(call!)
    }
    */
}


android/src/main/java/com/raccoonfink/plugins/eventsource/EventSource.java
package com.raccoonfink.plugins.eventsource;

import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.NativePlugin;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.launchdarkly.eventsource.*;
import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.concurrent.TimeUnit;
import okhttp3.OkHttpClient;
import okhttp3.Request;

enum ReadyState {
    CONNECTING,
    OPEN,
    CLOSED
}

@NativePlugin
public class EventSource extends Plugin implements EventHandler {

    private static final String TAG = "EventSource";

    private com.launchdarkly.eventsource.EventSource sse;
    private URL url;
    private boolean opened = false;

    @PluginMethod
    public void configure(final PluginCall call) {
        final String urlString = call.getString("url");

        if (urlString == null) {
            call.reject("you must provide a URL");
            return;
        }

        Log.i(TAG, "configuring event source with URL " + urlString);

        try {
            this.url = new URL(urlString);
        } catch (final MalformedURLException e) {
            Log.w(TAG, "invalid URL", e);
            call.reject("invalid URL");
        }

        if (this.sse != null) {
            Log.v(TAG, "configure() called, but there is an existing event source; making sure it's closed");
            this.opened = false;
            this.sse.close();
            this.sse = null;
        }

        final int reconnectTime = call.getInt("reconnectTime", 1000);
        final int maxReconnectTime = call.getInt("maxReconnectTime", 60000);
        final int backoffResetThreshold = call.getInt("backoffResetThreshold", 5000);
        final int idleTimeout = call.getInt("idleTimeout", 30000);

        final Request request = new Request.Builder().url(this.url).build();
        final OkHttpClient client = new OkHttpClient.Builder().readTimeout(idleTimeout, TimeUnit.MILLISECONDS).build();
        try {
            this.sse =
                new com.launchdarkly.eventsource.EventSource.Builder(this, this.url.toURI())
                    .reconnectTimeMs(reconnectTime)
                    .maxReconnectTimeMs(maxReconnectTime)
                    .backoffResetThresholdMs(backoffResetThreshold)
                    .readTimeoutMs(idleTimeout)
                    .build();
        } catch (final URISyntaxException e) {
            e.printStackTrace();
        }

        call.resolve();
    }

    private void safeNotifyListeners(final String eventName, final JSObject ret, final boolean retainUntilConsumed) {
        if (eventName == null || ret == null) {
            Log.w(TAG, "Failed to notify listeners, incomplete update");
            return;
        }
        try {
            this.notifyListeners(eventName, ret, retainUntilConsumed);
        } catch (final Exception e) {
            Log.e(TAG, "Failed to notify listeners, update dropped: " + ret.toString(), e);
        }
    }

    @PluginMethod
    public void open(final PluginCall call) {
        Log.i(TAG, "opening connection");

        final JSObject ret = new JSObject();
        ret.put("state", ReadyState.CONNECTING);
        this.safeNotifyListeners("readyStateChanged", ret, false);

        this.sse.start();
        this.opened = true;

        call.resolve();
    }

    @PluginMethod
    public void close(final PluginCall call) {
        Log.i(TAG, "closing connection");

        if (this.sse != null) {
            this.opened = false;
            this.sse.close();
            this.sse = null;
        }

        call.resolve();
    }

    public void onOpen() {
        if (!this.opened) {
            Log.v(TAG, "onOpen skipped (this.opened=false)");
            return;
        }

        Log.v(TAG, "onOpen");

        final JSObject ret = new JSObject();
        ret.put("value", null);
        this.safeNotifyListeners("open", ret, false);

        final JSObject rsRet = new JSObject();
        rsRet.put("state", ReadyState.OPEN);
        this.safeNotifyListeners("readyStateChanged", rsRet, false);
    }

    public void onMessage(final String event, final MessageEvent messageEvent) {
        if (!this.opened) {
            Log.v(TAG, "onMessage skipped (this.opened=false)");
            return;
        }

        Log.v(TAG, "onMessage: " + messageEvent.getData());

        final JSObject ret = new JSObject();
        ret.put("message", messageEvent.getData());
        this.safeNotifyListeners("message", ret, false);
    }

    public void onComment(final String comment) {
        Log.v(TAG, "onComment: " + comment);
    }

    public void onError(final Throwable throwable) {
        if (!this.opened) {
            Log.v(TAG, "onError skipped (this.opened=false)");
            return;
        }

        Log.e(TAG, "onError: " + throwable.getMessage(), throwable);

        final JSObject ret = new JSObject();
        ret.put("error", throwable.getMessage());
        this.safeNotifyListeners("error", ret, false);
    }

    public void onClosed() {
        if (!this.opened) {
            Log.v(TAG, "onClosed skipped (this.opened=false)");
            return;
        }

        Log.v(TAG, "onClosed");

        final JSObject ret = new JSObject();
        ret.put("state", ReadyState.CLOSED);
        this.safeNotifyListeners("readyStateChanged", ret, false);
    }
}
